crates/lsp/src/lsp.rs

```diff
@@ -33,7 +33,7 @@ use util::{ResultExt, TryFutureExt};
 
 const JSON_RPC_VERSION: &str = "2.0";
 const CONTENT_LEN_HEADER: &str = "Content-Length: ";
-const LSP_REQUEST_TIMEOUT: Duration = Duration::from_secs(60 * 2);
+const LSP_REQUEST_TIMEOUT: Duration = Duration::from_secs(60 * 4);
```

Comment:
Had to do this to give enough time for clojure-lsp to initialize

Reply:
Why does closure-lsp take that long in the first place though? Is that something that e.g. other IDEs using this LSP server suffer from as well?

Reply:
Most other editors don't set a timeout as far as I know. And yes, I was running the project in an "extreme" setting: my project has a ton of dependencies and I'm behind a VPN so downloading is a bit slow. Also, the project takes a while to compile.

But let me say I work at Nubank, which might be the biggest Clojure shop in the world, and most of our projects are similar to this. So I'd say it's pretty important for Clojure users.

Reply:
And yes, I have to wait a while for the lsp when opening a big project for the first time

Reply:
I mean, I do have to wait for rust-analyzer as well (see the attached screenshot), but it still sends a response to initialize request within 28ms. That's kind of my point; regardless of your project size, response to `Initialize` should ideally be pretty snappy.
![image](https://github.com/zed-industries/zed/assets/24362066/d510857e-642e-4388-a646-0331b49edfcc)
All I'm saying is that I don't think we should be bumping that timeout to 2 minutes just to accommodate for what (to me) looks like an anomaly rather than a norm. If you could point me to a pretty big open-source Closure project (whose size is somewhat representative of what you're dealing with) I could profile the `closure-lsp` with, that'd be golden.

Reply:
@prcastro when you run the build from this branch with `RUST_LOG=lsp=trace`, is there anything that stands out from the logs between us sending the initialise request and receiving a response to that request?
For example, here's my log for `RUST_LOG=lsp=trace cargo run -- ../random_projects/clojure_lsp`:
<details><summary>Details</summary>
<p>

[2024-01-29T21:46:11Z TRACE lsp] outgoing message:{"jsonrpc":"2.0","id":0,"method":"initialize","params":{"processId":null,"rootUri":"file:///","capabilities":{"workspace":{"workspaceEdit":{"documentChanges":true,"resourceOperations":["create","rename","delete"]},"didChangeConfiguration":{"dynamicRegistration":true},"didChangeWatchedFiles":{"dynamicRegistration":true,"relativePatternSupport":true},"symbol":{},"workspaceFolders":true,"configuration":true,"inlayHint":{"refreshSupport":true},"diagnostic":{}},"textDocument":{"completion":{"completionItem":{"snippetSupport":true,"resolveSupport":{"properties":["additionalTextEdits"]}},"completionList":{"itemDefaults":["commitCharacters","editRange","insertTextMode","data"]}},"hover":{"contentFormat":["markdown"]},"formatting":{},"onTypeFormatting":{},"definition":{"linkSupport":true},"codeAction":{"codeActionLiteralSupport":{"codeActionKind":{"valueSet":["refactor","quickfix","source"]}},"dataSupport":true,"resolveSupport":{"properties":["edit","command"]}},"rename":{"prepareSupport":true},"publishDiagnostics":{"relatedInformation":true},"inlayHint":{"dynamicRegistration":false,"resolveSupport":{"properties":["textEdits","tooltip","label.tooltip","label.location","label.command"]}},"diagnostic":{"relatedDocumentSupport":true}},"window":{"workDoneProgress":true},"experimental":{"serverStatusNotification":true}},"workspaceFolders":[{"uri":"file:///","name":""}]}}
[2024-01-29T21:46:11Z TRACE lsp] incoming message: {"jsonrpc":"2.0","method":"LogMessage","params":{"level":0,"message":"[DEBUG] [agent] [2024-01-29T21:46:11.717Z] Agent service starting","metadataStr":"[DEBUG] [agent] [2024-01-29T21:46:11.717Z]","extra":["Agent service starting"]}}
[2024-01-29T21:46:11Z INFO  lsp] 0 unhandled notification LogMessage:
    {
      "level": 0,
      "message": "[DEBUG] [agent] [2024-01-29T21:46:11.717Z] Agent service starting",
      "metadataStr": "[DEBUG] [agent] [2024-01-29T21:46:11.717Z]",
      "extra": [
        "Agent service starting"
      ]
    }
[2024-01-29T21:46:11Z TRACE lsp] incoming message: {"jsonrpc":"2.0","id":0,"method":"client/registerCapability","params":{"registrations":[{"id":"1148dea4-b570-478c-bd60-8a41da3242dc","method":"workspace/didChangeWorkspaceFolders","registerOptions":{}}]}}
[2024-01-29T21:46:11Z INFO  lsp] 0 unhandled notification client/registerCapability:
    {
      "registrations": [
        {
          "id": "1148dea4-b570-478c-bd60-8a41da3242dc",
          "method": "workspace/didChangeWorkspaceFolders",
          "registerOptions": {}
        }
      ]
    }
[2024-01-29T21:46:11Z TRACE lsp] incoming message: {"jsonrpc":"2.0","id":0,"result":{"capabilities":{"textDocumentSync":{"openClose":true,"change":2},"workspace":{"workspaceFolders":{"supported":true,"changeNotifications":true}}}}}
[2024-01-29T21:46:11Z TRACE lsp] Took 240.186334ms to receive response to "initialize" id 0

</p>
</details> 
You do not need to share the logs as I assume they might contain sensitive data, but maybe clojure-lsp has logs that could help us out here.

Reply:
@osiewicz I'm willing to revert the timeout to 2 minutes and then, on another pull request, make that configurable by the user. What do you think?

BTW, maybe you can test with this project: https://github.com/riemann/riemann

Reply:
> I'm willing to revert the timeout to 2 minutes and then, on another pull request, make that configurable by the user. What do you think?

I think this is a great idea, let's do that. Still, I think we should also get to the bottom of this and find a reason for a slow response to initialize. :p

It takes 200ms on my machine to get response to initialize request on that riemann repo you've shared.

Reply:
clojure-lsp really could take 1-2mins or more depending on the size of the project to respond to initialize request, it's on that request that analyze and scan the classpath and build its ASM, vscode, emacs and vim don't have timeouts IIRC on that request, I think would be nice to do the same as the LSP spec doesn't mention that

Reply:
Right, let's just add the knob to configure a timeout for initialize and call it a day. =)

Reply:
I reverted the timeout on this PR, and I'll soon start working on this configuration knob.